madsim/src/sim/task/mod.rs

```diff
@@ -234,13 +234,19 @@ impl Executor {
         };
         runnable.schedule();
 
+        let allow_system_thread =
+            crate::context::try_current(|h| h.allow_system_thread).unwrap_or_default();
         loop {
             self.run_all_ready();
             if task.is_finished() {
                 return task.now_or_never().unwrap();
             }
             let going = self.time.advance_to_next_event();
-            assert!(going, "no events, all tasks will block forever");
+            if allow_system_thread && !going {
+                std::thread::sleep(std::time::Duration::from_millis(1));
+            } else if !allow_system_thread {
+                assert!(going, "no events, all tasks will block forever");
+            }
```

Comment:
to make the logic clearer.

```suggestion
            if !going {
                if allow_system_thread {
                    // other system threads may wake up the tasks, wait for a while
                    std::thread::sleep(std::time::Duration::from_millis(1));
                } else {
                    panic!("no events, all tasks will block forever");
                }
            }
```